src/heartbeats.cpp

```diff
@@ -0,0 +1,605 @@
+    {
+        //TODO: consider use case: paused, if transfers are paused!?
+        return MegaBackupMonitor::State::ACTIVE;
+    }
+    else if (!sync->isEnabled())
+    {
+        return MegaBackupMonitor::State::DISABLED;
+    }
+    else
+    {
+        return MegaBackupMonitor::State::FAILED;
+    }
+}
+
+int MegaBackupMonitor::getHBSubstatus(MegaSync *sync)
+{
+    return sync->getError();
+}
+
+string MegaBackupMonitor::getHBExtraData(MegaSync *sync)
+{
+    return "";
+}
+
+BackupType MegaBackupMonitor::getHBType(MegaSync *sync)
+{
+    return BackupType::TWO_WAY; // TODO: get that from sync whenever others are supported
+}
+
+void MegaBackupMonitor::updateSyncInfo(handle backupId, MegaSync *sync)
+{
+
+    string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+    string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+    string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+    mClient->reqs.add(new CommandBackupPut(mClient, backupId, BackupType::TWO_WAY, sync->getMegaHandle(), localFolderEncrypted.c_str(),
+                                           deviceIDEncrypted.c_str(), nameEncrypted.c_str(),
+                                           getHBState(sync), getHBSubstatus(sync), getHBExtraData(sync).c_str()
+                                           ));
+}
+
+void MegaBackupMonitor::updateOrRegisterSync(MegaSync *sync)
+{
+    if (!sync)
+    {
+        return;
+    }
+    auto config = mClient->syncConfigs->get(sync->getTag());
+
+    handle syncID = UNDEF;
+    if (config)
+    {
+        syncID = config->getBackupId();
+    }
+
+    bool pushInPending = false;
+
+    string extraData;
+    if (syncID == UNDEF) //register
+    {
+
+        string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+        string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+        string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+        if (std::find(mPendingBackupPuts.begin(), mPendingBackupPuts.end(), sync->getTag()) == mPendingBackupPuts.end())
+        {
+            pushInPending = true;
+            mClient->reqs.add(new CommandBackupPut(mClient, BackupType::TWO_WAY, sync->getMegaHandle(), localFolderEncrypted.c_str(),
+                                           deviceIDEncrypted.c_str(), nameEncrypted.c_str(),
+                                           getHBState(sync), getHBSubstatus(sync), getHBExtraData(sync)
+                                           ));
+        }
+        else // ID not received yet, let's queue the update (copying sync data)
+        {
+            LOG_debug << " Queuing sync update, register is on progress for sync: " << sync->getLocalFolder();
+            mPendingSyncUpdates[sync->getTag()].reset(sync->copy()); // we replace any previous pending updates
```

Comment:
Instead of a `MegaSync` instance, we could maintain a copy of the data that requires to be updated, ie. the backup type, the local folder, the name of the backup...
That would make it work with CU indifferently :)

Reply:
`updateSyncInfo` is called indiferently from `updateOrRegisterSync` and from the code that handles these mPendingSyncUpdates.
If we were to solve this elegantly we should create some BackupInfo class with a (const MegaSync &sync, handle Id = UNDEF) constructor, and refactor the code to update/register.
Note to self: updateOrRegisterSync is accepting MegaSync, we would need to move the code that looks for cached backup Id outside ...

Reply:
Implemented here: https://github.com/meganz/sdk/pull/2140, however, we cannot have a mPendingSyncUpdates generic for all types of backups, since it needs to be a map with sync tag as key, so as to have one only update, and we need to be able to find it in registration callback. The code in that PR is a little bit more prepared for CU backups, but we will need to keep a separate map of `mPendingCUUpdates`, and some code similar to void `updateOrRegisterSync(MegaSync *sync)` but for CUs.